Initial version of the web interface

alexandru-elisei opened the pull request:
This is just a prototype, I know it looks ugly at places, this is just the initial version to show you how we can generate the web page automatically based on the templates needed to create a plugin. If you agree with it I'll modify it to make it look better.

I've tested the web interface, right now I can:
- generate a plugin, by specifying values for all the variables by hand or by load a recipe file
- save the recipe constructed from the form variables.

I've created a plugin from a saved recipe, so that works fine.

There's also some fixes to other files.

While you revise this commit I'll work on creating the get_template_variables() function for each skel type, no matter what we decide about the web interface we're still going to need a function like that.

alexandru-elisei commented on classes/local/skel/version_php_file.php:
```diff
@@ -37,6 +37,20 @@
 class version_php_file extends php_internal_file {
 
+    /** @var string[] Moodle versions. */
+    protected static $moodleversions = [
```
Made the attribute static so I can send use it in the static function get_template_variables as a list of accepted values for the variable required.

alexandru-elisei commented on classes/local/util/manager.php:
```diff
@@ -67,9 +67,42 @@ public static function instance($logger) {
+     * @param string $component.
+     * @return string[].
+     */
+    public static function get_component_variables($component) {
```
This is not complete yet, I'll add code here as I write the get_template_variables() function for the skel classes.

For each skel class that represents a template which requires variables to be present in the recipe there'll be a call like this one:
```php
$recipevariables = \tool_pluginskel\local\skel\skelclass::get_template_variables($component);
```

alexandru-elisei commented on classes/local/util/manager.php:
```diff
@@ -748,7 +781,7 @@ protected function prepare_lang_file() {
         $this->init_file('lang/en/'.$this->recipe['component'].'.php', 'lang', [
```
The prepare_lang_file() function will be deleted in a future patch, it isn't used anywhere.

alexandru-elisei commented on classes/step1_form.php:
```diff
@@ -0,0 +1,220 @@
+            $currentcount = $currentcount + 1;
+        }
+
+        $mform->addElement('submit', 'addmore_'.$name, get_string('addmore', 'tool_pluginskel'));
```
I haven't implemented this yet. When the user presses the Add more button more text elements will be created on the web page.

The best way to do it would be with Javascript, I'll find a way to do it.

alexandru-elisei commented on index.php:
```diff
@@ -0,0 +1,170 @@
+                    } else {
+                        $count = count($recipe[$var['name']]);
+                    }
+                    $data[$formvariable] = $count;
```
I need this for variables that have an array as value (like dependencies). I don't know how many elements are in the array, so I need to save their number when creating the form.

mudrd8mz commented on classes/local/util/yaml.php:
```diff
@@ -45,7 +45,7 @@ class yaml {
      * @return array
      */
-    public static function decode_string($yaml) {
+    public static function decode_string($string) {
```
This should be renamed in the @param section above, too.

alexandru-elisei commented:
I've rebased against master.

I've also modified an error that I introduced when modifying the way step1_form is generated, I think you will need to clone the branch with these changes for the web interface to work properly.

alexandru-elisei commented on classes/step1_form.php:
```diff
@@ -0,0 +1,284 @@
+                $this->add_advcheckbox($variable, $recipe, true);
+            }
+
+            if (!empty($variable['values'])) {
```
I don't really like this. I think I should use a hint instead of relying on having the values declared for creating a select element.

Maybe the hint should be something like 'multiple-choice'.

alexandru-elisei commented on classes/local/util/manager.php:
```diff
@@ -67,9 +67,64 @@ public static function instance($logger) {
+        $featurevars = array();
+
+        // Adding strings.
+        $featurevars[] = array('name' => 'strings', 'hint' => 'multiple', 'values' => array('id', 'text'));
```
I don't like this at all. I will change it.

I think the best way to do it would be like this:

```php
array('name'=>'string', 'hint' => 'array', values => array(
    array('name'=>'id', 'hint' => 'boolean', 'required'=>false), // This is just an example.
    array('name'=>'text')
);
```
Each value from the 'values' array would be a full template variable unto itself.

mudrd8mz commented:
Hi Alex. Sorry I did not have a chance to look at you recent changes. I am leaving tomorrow and will be out of office for the next week. I will check the work in progress once I am back on 8 August.

alexandru-elisei commented:
Sure, no problem, after all it's still a work in progress.

About the travis failed check, I'm getting a DOM node value and transforming it into a number like this:
```javascript
var count = +$('#id').val();
```
Travis throws an error because I have no space between the '+' and the '$'. This is the coding style that I saw recommended on the jquery website.

Should I change how I wrote that?

mudrd8mz commented:
I haven't seen this syntax used before, what is it good for?

alexandru-elisei commented:
The string returned by .val() is automatically converted into a number.

alexandru-elisei commented:
This is a final version.

I have implemented all the features except adding the index in front of array variables. I will work on it while you take a look at the code.

There is a lot of code and I've tried to make it as simple and easy to follow as possible, but I'm sure there'll be some things that I'll need to rewrite.

There is one limitation on the recipe variables: there cannot be more than two levels of array nesting. For example:
```
capabilities:
    - name: cap
      ...
    - archetypes:
        - role: student
        - permission: CAP_ALLOW
``` 

It has two levels of nesting (`archetypes` represents the second level of nesting).

So is `backup_moodle2`:
```
mod_features:
    backup_moodle2:
        restore_task: true
        ...
        backup_elements:
            - name: first
            - name: second
```

In this case, `backup_moodle2` is an associative array which has a numerically indexed array (`backup_elements`) as one of the values.

Implementing more than two levels of nesting seems useless at this time (we don't have a recipe configuration that requires that).

Also, I haven't implemented two levels of nesting for associative arrays. We don't have a recipe option that looks like that, and we can express two associative arrays nested inside each other as one big associative array if we need to.

mudrd8mz commented on amd/src/addmore.js:
```diff
@@ -0,0 +1,433 @@
+                prevIndex = prevIndex + '. ';
+                newIndex = newIndex + '. ';
+                newElementHtml = newElementHtml.replace('label for="'+newElementId+'">'+prevIndex,
+                                                        'label for="'+newElementId+'">'+newIndex);
```
Please see failed Travis for missing spaces here.

alexandru-elisei replied:
Sorry, forgot to check Travis after pushing the commit. Fixed.

mudrd8mz commented on classes/step0_form.php:
```diff
@@ -0,0 +1,71 @@
+        $mform->addElement('select', 'componenttype', get_string('skelcomponenttype', 'tool_pluginskel'), $plugintypes);
+
+        $mform->addElement('text', 'componentname', get_string('skelcomponentname', 'tool_pluginskel'), null);
+        $mform->setType('componentname', PARAM_TEXT);
```
This should probably better be PARAM_PLUGIN.

alexandru-elisei replied:
Sure.

mudrd8mz commented on index.php:
```diff
@@ -0,0 +1,418 @@
+ * @param string $countprefix The prefix for the variable that will hold the number of values for the variables.
+ * @return string[]
+ */
+function get_array_variable_count_from_recipe($templatevars, $recipe, $countprefix = '') {
```
It is quite ugly and bad practise to have functions defined in the main action scripts like this. Introduce a helper class for the web interface, or at least move it to locallib.php or so.

alexandru-elisei replied:
Sure, I can create a helper class, I was thinking the same thing (those global functions do look out of place).

mudrd8mz commented on settings.php:
```diff
@@ -0,0 +1,36 @@
+
+// Entry in Site administration -> Development -> Generate plugin skeleton.
+if ($hassiteconfig) {
+    $ADMIN->add(
```
Can we have a global setting for this tool that would allow to define default values that are likely to be re-used every time. Typically I would like to fill my name and copyright values so that I do not need to retype them all the time.

mudrd8mz commented:
Thanks Alex for this. I generally like the implementation and I am going to merge it now. Beside the comments left in the code (see above), there are two major things that I think should be addressed yet:

* I remember we discussed it. I really don't think that `hint` is a good name for the thing in the patch. It should really read `type`. Intuitively, the `hint` would contain a human readable help (hint) text for the user on what to fill the form with. Let me suggest to mass search and replace it yet.
* We really need help icons supported. I do not think you necessarily need to prepare all the help strings. But the mechanism should be in the place so that it is clear on how we can add remaining help strings.
* Please note that for checkbox fields, we could provide a checkbox title that could be displayed next to it (and _that_ would be what I call a hint). AFAIK other moodleform fields do not support it unfortunately.

Code-wise, the plugin is cool. We just need to improve the UI a bit (in terms of intuitiveness and built-in help) as that is the first impression that most people will have.

alexandru-elisei commented:
- Ok, I can rename `hint` to `type`.
- As for help icons, absolutely, we need them. I was waiting for your review for implementing them. 

Because some fields are not unique (for example, I am using `name` for a lot of variable options), an unique name for the help icons is needed. I can use their position in the recipe to create an unique name for the help string.

For example:
```
mobile_addons:
    - name: 'my_first_addon'
      dependencies:
          - name: dependency_one
```
The help string for the `my_first_addon` name would be `mobile_addons_name_help`, and for the first depenedency name would be `mobile_addons_dependencies_name_help`. These would be passed to the `add_<mform elemens>()` functions as an argument.

mudrd8mz commented:
> Because some fields are not unique (for example, I am using name for a lot of variable options)

Let me discourage from that. We learnt that it is better to have explicit own string for every field even if in English they read the same. For example in Czech, we have different translation of "Name" depending on whether it is name of a thing or a name of a person.

So I would suggest to have explicit strings like `mobile_addons_name` and `mobile_addons_dependencies_name` even if they both are defined as "Name" in the English pack. 

At the end, you are likely to realise that having them all defined more explicitly like "Capability name", "Addon name", "Dependency name" etc. is actually better than simple "Name" everywhere.
